### PR TITLE
#7105 - Added reduced immunizations to the frontend

### LINKED ISSUE #7105: [Immunizations] Reduced entry option for vaccinations [3]
### Situation Description

With the implementation of the immunization module, vaccination information was migrated out of the case, contact and event participant to the immunization entity of the person. However, if users don't use the extensive immunization module, vaccination information would be hidden and there is also no easy way to document vaccinations for a specific disease.

In order to make it easy to add the vaccination information if the immunization module is inactive, a reduced version should be added as an alternative option.

Prerequisites: #7111, #7148

### Feature Description

When the immunization module is inactive, 
- add a vaccination card to the person entity and it's related entities (cases, contacts, event participants) dependending on the disease
    - Heading: "Vaccinations"
    - Creation button: "New Vaccination"
    - Card content: Display each entry separated by a line; content in the card should include the `vaccineName` and the `vaccinationDate`; in persons, it should also include the `disease`
        - each entry should have an edit icon to open and edit the respective entry. Closing the pop-up will let the user stay on the previous form

- Upon clicking "New vaccination", open the existing vaccination pop-up (cf. new entry for table in immunization) but exclude the pregnancy and pre-existing conditions fields.
    - saving the entry should leave the user on the previous form
    - if the user creates a new vaccination out of the person entity, the disease field also needs to be displayed which is usually part of the immunization entity. This is necessary in order to display the vaccinations correctly across entities.
    - display all existing vaccinations in each entity of the person but grey out the ones that are not relevant for the respective entity; determine relevance as follows
        - cases: `vaccinationDate` before `onsetDate`, if not given `reportDate`
        - contacts: `vaccinationDate` before `lastContactDate`, if not given `reportDate`
        - event participants: `vaccinationDate` before event `startDate`, if not given event `endDate` (if before report date), if not given event `reportDate`
    - if the vaccination is displayed as not relevant for the respective entity, add in info icon to the entry
        - upon hover, show the following message: _"This persons vaccination is not relevant for this case/contact/event participant."_

- If a relevant vaccination exists or a new relevant vaccination is added, set the `vaccinationStatus` to `VACCINATED`

### Additional Information

Vaccinations Card

In cases, contacts, event participants
New vaccination entry

### PR DESCRIPTION
**Pull request by MateStrysewske**
Fixes #7105

### CODE REVIEW

File: sormas-api/src/main/java/de/symeda/sormas/api/vaccination/VaccinationFacade.java [hunk 1]
```diff
@@ -40,7 +40,13 @@ VaccinationDto create(
 
 	List<VaccinationDto> getAllVaccinations(String personUuid, Disease disease);
 
+	List<VaccinationListEntryDto> getEntriesList(String personUuid, Disease disease, Integer first, Integer max);
```
Comment (MartinWahnschaffe) [inline sormas-api/src/main/java/de/symeda/sormas/api/vaccination/VaccinationFacade.java:43]: Add a comment to the create method that explain how it automatically assigns the vaccination to an immunization.

File: sormas-ui/src/main/java/de/symeda/sormas/ui/vaccination/VaccinationController.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+		final CommitDiscardWrapperComponent<VaccinationEditForm> cdwComponent =
+			new CommitDiscardWrapperComponent<>(form, UserProvider.getCurrent().hasUserRight(UserRight.IMMUNIZATION_CREATE), form.getFieldGroup());
+		cdwComponent.getCommitButton().setCaption(doSave ? I18nProperties.getCaption(Captions.actionSave) : I18nProperties.getString(Strings.done));
+
+		VaadinUiUtil.showModalPopupWindow(cdwComponent, I18nProperties.getCaption(VaccinationDto.I18N_PREFIX));
+
+		cdwComponent.addCommitListener(() -> {
+			if (!form.getFieldGroup().isModified()) {
+				if (doSave && immunization != null) {
+					FacadeProvider.getVaccinationFacade().save(form.getValue());
```
Comment (MartinWahnschaffe) [inline sormas-ui/src/main/java/de/symeda/sormas/ui/vaccination/VaccinationController.java:74]: The naming for save vs. create in the facade makes this quite confusing. Let's discuss tomorrow.

File: sormas-ui/src/main/java/de/symeda/sormas/ui/vaccination/VaccinationController.java [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+	public void create(
+		ImmunizationReferenceDto immunization,
+		Disease disease,
+		UiFieldAccessCheckers fieldAccessCheckers,
+		Consumer<VaccinationDto> commitCallback) {
+		create(immunization, null, null, null, disease, fieldAccessCheckers, false, commitCallback);
+	}
+
+	public void create(
```
Comment (MartinWahnschaffe) [inline sormas-ui/src/main/java/de/symeda/sormas/ui/vaccination/VaccinationController.java:48]: Add a comment that explains how either the immunization or the other fields are optional.

OR: Make this method private and add another public create method for the other use case.

**Review by MartinWahnschaffe — CHANGES_REQUESTED** [review]
The VaccinationFacade.create method will currently take an immunization. From my understanding it should not assign the vaccination to an existing immunization that is not VACCINATION or VACCINATION_RECOVERY. I'd suggest to also add a unit test for this scenario.

File: sormas-backend/src/main/java/de/symeda/sormas/backend/vaccination/VaccinationFacadeEjb.java [hunk 4]
```diff
@@ -259,6 +270,22 @@ protected void updateVaccinationStatuses(Date newVaccinationDate, Date currentVa
...
 
+	@Override
+	public void delete(String uuid) {
+		if (!userService.hasRight(UserRight.IMMUNIZATION_DELETE)) {
+			throw new UnsupportedOperationException("User " + userService.getCurrentUser().getUuid() + " is not allowed to delete vaccinations");
+		}
+
+		Vaccination vaccination = vaccinationService.getByUuid(uuid);
+		vaccination.getImmunization().getVaccinations().remove(vaccination);
+		immunizationService.ensurePersisted(vaccination.getImmunization());
```
[the remark this answers is not in the document]
Comment (MateStrysewske) [reply sormas-backend/src/main/java/de/symeda/sormas/backend/vaccination/VaccinationFacadeEjb.java]: Additionally rename the method to `deleteWithImmunization`

File: sormas-ui/src/main/java/de/symeda/sormas/ui/vaccination/VaccinationController.java (hunk 2 above)
Comment (MateStrysewske) [reply sormas-ui/src/main/java/de/symeda/sormas/ui/vaccination/VaccinationController.java:74]: Rename the `create` method to `createWithImmunization`

**Review by MartinWahnschaffe — CHANGES_REQUESTED** [review]
Change date of a case/contact/? is not updated when it's immunization status is updated.

**Pull request merged**